crates/log-server/src/rocksdb_logstore/builder.rs

```diff
@@ -0,0 +1,195 @@
+
+        let db_spec = DbSpecBuilder::new(db_name, data_dir, db_options(options))
+            .add_cf_pattern(
+                CfExactPattern::new(DATA_CF),
+                cf_data_options(options.rocksdb_memory_budget()),
+            )
+            .add_cf_pattern(
+                CfExactPattern::new(METADATA_CF),
+                cf_metadata_options(options.rocksdb_memory_budget()),
+            )
+            // not very important but it's to reduce the number of merges by flushing.
+            // it's also a small cf so it should be quick.
+            .add_to_flush_on_shutdown(CfExactPattern::new(METADATA_CF))
+            .ensure_column_families(cfs)
+            .build()
+            .expect("valid spec");
+        let db_name = db_spec.name().clone();
+        // todo: use the returned rocksdb object when open_db returns Arc<RocksDb>
+        let _ = db_manager
+            .open_db(updateable_rocksdb_options, db_spec)
+            .await?;
+        let rocksdb = db_manager.get_db(db_name).unwrap();
+
+        Ok(Self {
+            rocksdb,
+            updateable_options,
+        })
+    }
+
+    pub async fn start(self, task_center: &TaskCenter) -> Result<RocksDbLogStore, ShutdownError> {
+        let RocksDbLogStoreBuilder {
+            rocksdb,
+            updateable_options,
+        } = self;
+        // todo (asoli) load up our loglet metadata cache.
+        let writer_handle =
+            LogStoreWriter::new(rocksdb.clone(), updateable_options.clone()).start(task_center)?;
+
+        Ok(RocksDbLogStore {
+            updateable_options,
+            rocksdb,
+            writer_handle,
+        })
+    }
+}
+
+fn db_options(options: &LogServerOptions) -> rocksdb::Options {
+    let mut opts = rocksdb::Options::default();
+
+    // enable atomic flushes to not persist inconsistent data in case WAL
+    // is disabled
+    if options.rocksdb.rocksdb_disable_wal() {
+        opts.set_atomic_flush(true);
+    }
+
+    // This is Rocksdb's default, it's added here for clarity.
+    //
+    // Rationale: If WAL tail is corrupted, it's likely that it has failed during write, that said,
+    // we can use absolute consistency but on a single-node setup, we don't have a way to recover
+    // from it, so it's not useful for us. Even for a replicated loglet setup, the recovery is
+    // likely to depend on the last durably committed record for a quorum of nodes in the write
+    // set.
+    opts.set_wal_recovery_mode(rocksdb::DBRecoveryMode::TolerateCorruptedTailRecords);
+
+    opts
+}
+
+fn cf_data_options(
+    memory_budget: usize,
+) -> impl Fn(rocksdb::Options) -> rocksdb::Options + Send + Sync + 'static {
+    move |mut opts| {
+        // memory budget is in bytes. We divide the budget between the data cf and metadata cf.
+        // data 10% to metadata 90% to data.
+        let memtables_budget = (memory_budget as f64 * DATA_CF_BUDGET_RATIO).floor() as usize;
```

Comment:
Is `DATA_CF_BUDGET_RATIO` the ratio? If yes, then it is currently set to 85%. Maybe we don't specify the ratio in the comment.